From 5c7ae01d88657a5cfae9a5b80d62404778b66b8f Mon Sep 17 00:00:00 2001
From: Sammy He <r62914@freescale.com>
Date: Thu, 4 Nov 2010 00:38:26 +0800
Subject: [PATCH 191/442] ENGR00125671 Obvious serration for the interlaced
 streams

Not set correct field for interlaced MPEG4 stream. Add this process
in test app.

Signed-off-by: Sammy He <r62914@freescale.com>
---
 test/mxc_vpu_test/dec.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/test/mxc_vpu_test/dec.c b/test/mxc_vpu_test/dec.c
index 66708c0..1772f68 100644
--- a/test/mxc_vpu_test/dec.c
+++ b/test/mxc_vpu_test/dec.c
@@ -32,6 +32,8 @@ static FILE *fpSliceBndLogfile = NULL;
 static FILE *fpMvLogfile = NULL;
 static FILE *fpUserDataLogfile = NULL;
 
+static int isInterlacedMPEG4 = 0;
+
 #define FN_FRAME_BUFFER_STATUS "dec_frame_buf_status.log"
 #define FN_ERR_MAP_DATA "dec_error_map.log"
 #define FN_QP_DATA "dec_qp.log"
@@ -941,8 +943,10 @@ decoder_start(struct decode *dec)
 				}
 				if (outinfo.vc1_repeatFrame)
 					info_msg("dec_idx %d : VC1 RPTFRM [%1d]\n", decIndex, outinfo.vc1_repeatFrame);
-			} else if (dec->cmdl->format == STD_AVC) {
-				if (outinfo.interlacedFrame) {
+			} else if ((dec->cmdl->format == STD_AVC) ||
+				   (dec->cmdl->format == STD_MPEG4)) {
+				if ((outinfo.interlacedFrame) ||
+				    ((dec->cmdl->format == STD_MPEG4) && isInterlacedMPEG4)) {
 					if (outinfo.topFieldFirst)
 						field = V4L2_FIELD_INTERLACED_TB;
 					else
@@ -1584,6 +1588,8 @@ decoder_parse(struct decode *dec)
 					break;
 				}
 			}
+			isInterlacedMPEG4 = initinfo.interlace;
+
 			info_msg("Mpeg4 Profile: %d Level: %d Interlaced: %d\n",
 				profile, level, initinfo.interlace);
 			/*
-- 
1.8.0

